### PR TITLE
refactor: use pydantic

### LINKED ISSUE #2: Explore more efficient serialization libraries
https://github.com/rossmacarthur/serde

### PR DESCRIPTION
**Pull request by fubuloubu**
fixes: #2

This major refactor uses the venerable pydantic library to clean up a lot of the custom serialization and validation code using the more standardized pydantic approach.

There was some difficulties and changes necessary to make use of this newer library, but it seems to function quite well and caught a lot of latent data validation errors that were not getting caught.

### CODE REVIEW

File: tokenlists/typing.py [hunk 1]
```diff
@@ -1,92 +1,133 @@
...
+
+    @validator("extensions")
+    def extensions_must_contain_simple_types(cls, d: Optional[dict]) -> Optional[dict]:
+        if not d:
+            return d
 
+        # `extensions` is `Dict[str, Union[str, int, bool, None]]`, but pydantic mutates entries
+        for val in d.values():
+            if not isinstance(val, (str, int, bool)) and val is not None:
+                raise ValueError("")
```
Comment (lost-theory) [inline tokenlists/typing.py]: Is there supposed to be an error message here?

Comment (fubuloubu) [reply tokenlists/typing.py]: ah yes good catch lol
```suggestion
                raise ValueError(f"Incorrect extension field value: {val}")
```

File: tokenlists/typing.py [hunk 2]
```diff
@@ -1,92 +1,133 @@
...
-        if self.tags is None:
-            del data["tags"]
-        if self.logoURI is None or self.logoURI == "":
-            del data["logoURI"]
+    logoURI: Optional[AnyUrl] = None
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        all_tags = set(tag for token in self.tokens for tag in (token.tags if token.tags else []))
```
Comment (lost-theory) [inline tokenlists/typing.py]: This is hard for me to parse. How about:

```
tags_lists = [token.tags or [] for token in self.tokens]
all_tags = reduce(tags_lists, [])
```

Not much better for readability but it breaks up the nested list comprehension.

**Review by lost-theory — APPROVED** [review]
Looks good to me.

File: tokenlists/typing.py (hunk 2 above)
Comment (fubuloubu) [reply tokenlists/typing.py]: fixed in e156e10

**Pull request merged**